lxd/profiles_utils.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+func doProfileUpdate(d *Daemon, name string, id int64, profile *api.Profile, req api.ProfilePut) Response {
+	// Sanity checks
+	err := containerValidConfig(d, req.Config, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	err = containerValidDevices(req.Devices, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	containers := getContainersWithProfile(d, name)
+
+	// Check if the root device is supposed to be changed or removed.
+	oldProfileRootDiskDevice := containerGetRootDiskDevice(profile.Devices)
+	newProfileRootDiskDevice := containerGetRootDiskDevice(req.Devices)
+	if len(containers) > 0 && oldProfileRootDiskDevice != nil && oldProfileRootDiskDevice["pool"] != "" {
+		if newProfileRootDiskDevice == nil || newProfileRootDiskDevice["pool"] == "" || (oldProfileRootDiskDevice["pool"] != newProfileRootDiskDevice["pool"]) {
+			// Check for any containers that have a local root disk device.
+			for i := 0; i < len(containers); i++ {
+				for _, v := range containers[i].LocalDevices() {
+					if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" && v["pool"] != "" {
+						copy(containers[i:], containers[i+1:])
+						containers[len(containers)-1] = nil
+						containers = containers[:len(containers)-1]
+						i -= 1
+						break
+					}
+				}
+			}
+
+			// Check if another profile provides the same root disk device.
+			for i := 0; i < len(containers); i++ {
+				for _, p := range containers[i].Profiles() {
+					// Search through all of the containers
+					// profiles. Only skip the one we
+					// are about to change.
+					if p == profile.Name {
+						continue
+					}
+
+					_, pInfo, err := dbProfileGet(d.db, p)
+					if err != nil {
+						continue
+					}
+
+					for _, v := range pInfo.Devices {
+						// There's another profile that provides a root device on the same pool.
+						// So we can delete the container from the list.
+						if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" && oldProfileRootDiskDevice["pool"] == v["pool"] {
+							copy(containers[i:], containers[i+1:])
+							containers[len(containers)-1] = nil
+							containers = containers[:len(containers)-1]
+							i -= 1
+							break
+						}
+					}
+				}
+			}
+			if len(containers) > 0 {
+				return BadRequest(fmt.Errorf("At least one container relies on this profiles' root disk device."))
+			}
+		}
+	}
```

Comment:
So the goal is to know whether any container relies on this profile's root device and if that's the case, prevent changes to it.

```
// Check for containers using the device
for _, container := range containers {
    // Check if the device is locally overridden 
    _, ok := container.Devices[device_name]
    if ok {
        continue
    }

    // Check what profile the device comes from
    profiles = container.Profiles()
    for i := len(profiles)-1; i >= 0; i-- {
        _, profile, err := dbProfileGet(d.db, profiles[i])
        if err != nil {
            return err
        }

        // Found the profile
        _, ok := profile.Devices[device_name]
        if ok {
            if profiles[i] == profile_name {
                return fmt.Errorf("At least one container relies on this profile's root disk device.")
            } else {
                break
            }
        }
    }
}
```

That should achieve the same goal and seems a bit more readable to me.